src/playlist.rs

```diff
@@ -176,7 +176,12 @@ impl VariantStream {
             self.write_stream_inf_common_attributes(w)?;
             write_some_attribute_quoted!(w, ",AUDIO", &self.audio)?;
             write_some_attribute_quoted!(w, ",SUBTITLES", &self.subtitles)?;
-            write_some_attribute_quoted!(w, ",CLOSED-CAPTIONS", &self.closed_captions)?;
+            // handle `CLOSED-CAPTURES=NONE` case
+            if self.closed_captions.as_deref().eq(&Some("NONE")) {
```

Comment:
This seems a bit ugly. Maybe this should be an enum like

```rust
enum ClosedCaptionGroupId {
    None,
    GroupId(String),
}
```

How exactly is the behaviour different between having no `CLOSED-CAPTION` attribute and having it set to `NONE`?

Reply:
That would be nice, but it is not the same thing. 

This is from the standard says (also quoted in the issue #44 )

```
      If the value is the enumerated-string value NONE, all EXT-X-
      STREAM-INF tags MUST have this attribute with a value of NONE,
      indicating that there are no closed captions in any Variant Stream
      in the Multivariant Playlist.  Having closed captions in one
      Variant Stream but not another can trigger playback
      inconsistencies.
```

If I remember correctly, `CLOSED-CAPTION=NONE` means players will not offer any closed caption choice, even if received as CEA 608/708. When the attribute not specified, user should still have the choice to use those. I believe that is exactly what they're doing in the manifest (see the named issue) - they're suppressing any closed caption that might be received in the stream and only allowing specified subtitles.

Reply:
Right, that makes sense. So it really has a different semantic meaning than no CC group id at all.

In that case I think such an enum would make most sense because otherwise this is rather stringly-typed and you can't distinguish `NONE` from `"NONE"`.

Reply:
The enum does make sense, sure. The issue is that it would require a much bigger change on the parser side because today it's just ignoring double quotes ingesting a value to all the kv pairs.

(url=https://github.com/rutgersc/m3u8-rs/blob/master/src/parser.rs#L615)

Reply:
You would change that then to

```
alt((tag("NONE").map(|_| ClosedCaptionGroupId::None), quoted.map(|q| ClosedCaptionGroupId::String(q))),
```

or similar

Reply:
I.e. other unquoted strings are AFAIU not allowed

Reply:
If other unquoted strings are allowed, you could add a third variant to the `alt` that is with `unquoted`

Reply:
I did look and didn't see any other mixes of double-quoted and enumerated-strings, other than `CLOSED-CAPTIONS` and the `NONE` is the only value. In other cases it's usually either one or another. So, this should be safe

Nice `nom`-ing there by the way. I haven't seen anything beyond ver 4, when they only started coming up with functions. I'll look at this tomorrow, it's 2:27am here :)

Reply:
Sure, take your time and thanks for looking into this :)

Reply:
wait, this won't work as other attributes can have enumerated-strings `NONE` value.

* `HDCP-LEVEL` for `EXT-X-STREAM-INF`
* `METHOD` for `EXT-X-KEY`

These attributes probably are not rendered out correctly either. Something is to check in a future PR.

At any rate, it has to be specifically for `CLOSED-CAPTIONS` either when manifests are parsed or rendered. Since we have one-fit-all approach on the parser side for all attributes, the thought was that it's easier to identify when a manifest is rendered for each specific case.

Reply:
Yes that's correct. The current parsing of those two at least is also wrong. If you don't fix it as part of this PR, can you create an issue about that? :)

> the thought was that it's easier to identify when a manifest is rendered for each specific case.

I think you will need to special-case that in 3 places
 1. After parsing when mapping the known attributes to the fields in the structs
 2. Having specific types in the structs for handling these special enumerated values and I guess for unknown attributes we should distinguish between quoted and unquoted too for avoiding information loss. For the known attributes the current `String` value should probably also distinguish between quoted/unquoted in addition to the known enumerated values. E.g. `enum ClosedCaptionGroupId { None, String(String), UnquotedString(String), }` or what do you think?
 3. When writing things out

Reply:
Yo guys, thanks for looking into this! It seems clear that we need to be able to distinguish between quoted and unquoted values, and likely support the enumerated values that are defined by the standard with specific types like you suggested.

So ignore this if i'm completely wrong on this, its been quite a while for me, but would it then be enough to:
1. Change or add [this](https://github.com/rutgersc/m3u8-rs/blob/master/src/parser.rs#L615) signature to something like `fn key_value_pair(i: &[u8]) -> IResult<&[u8], (String, QuotedOrUnquoted)>`? Then all the existing code can still work the same, and the enumerated values can be translated to the specific types, like `ClosedCaptionGroupId` and `HdcpLevel`.
2. Have the types that the standard defines enumerations for, also make use of this `QuotedOrUnquoted` type, if it makes sense for the type? So it could be `enum ClosedCaptionGroupId { None, QuotedOrUnquoted(String) }`. Actually, either would be fine right?

This does look like a bigger change then, unfortunately.

Reply:
Yes that's basically what I meant :)

Reply:
Yes, that's what I've realized too after giving it some more thoughts... I think it's a bit more changes than this, but it seems to be a correct path forward.

Reply:
Okay, before I rewrite the whole thing, could somebody make sure I'm not overcomplicating this?

So, adding the enum:

```
pub enum QuotedOrNone {
    None,
    Quoted(String),
}
```

Then changing the signature 
```
fn key_value_pair(i: &[u8]) -> IResult<&[u8], (String, QuotedOrNone)> {
```

Then, well have to return correct value:

```
fn key_value_pair(i: &[u8]) -> IResult<&[u8], (String, QuotedOrNone)> {
    map(
        tuple((
            peek(none_of("\r\n")),
            map_res(take_until("="), from_utf8_slice),
            char('='),
            alt((quoted, unquoted)),
            opt(char(',')),
        )),
        |(_, left, _, right, _)| { match right.as_str() {
            "NONE" => (left, QuotedOrNone::None),
            _ => (left, QuotedOrNone::Quoted(right))
        } } ,
    )(i)
}
```

Then we have type mismatch in `key_value_pairs` so, doing this:

```
fn key_value_pairs(i: &[u8]) -> IResult<&[u8], HashMap<String, QuotedOrNone>> {
```

Then `from_hashmap()` is has type mismatch because of `VariantStream` enum, so changing this:

```pub struct VariantStream {
    pub is_i_frame: bool,
    pub uri: String,

    // <attribute-list>
    pub bandwidth: QuotedOrNone,
    pub average_bandwidth: Option<QuotedOrNone>,
    pub codecs: Option<QuotedOrNone,
    pub resolution: Option<QuotedOrNone>,
    pub frame_rate: Option<QuotedOrNone>,
    pub hdcp_level: Option<QuotedOrNone>,
    pub audio: Option<QuotedOrNone>,
    pub video: Option<QuotedOrNone>,
    pub subtitles: Option<QuotedOrNone>,
    pub closed_captions: Option<QuotedOrNone>,
    // PROGRAM-ID tag was removed in protocol version 6
}

impl VariantStream {
    pub fn from_hashmap(mut attrs: HashMap<String, QuotedOrNone>, is_i_frame: bool) -> VariantStream {
        VariantStream {
            is_i_frame,
            uri: attrs.remove("URI").unwrap_or_default(),
            bandwidth: attrs.remove("BANDWIDTH").unwrap_or_default(),
            average_bandwidth: attrs.remove("AVERAGE-BANDWIDTH"),
            codecs: attrs.remove("CODECS"),
            resolution: attrs.remove("RESOLUTION"),
            frame_rate: attrs.remove("FRAME-RATE"),
            hdcp_level: attrs.remove("HDCP-LEVEL"),
            audio: attrs.remove("AUDIO"),
            video: attrs.remove("VIDEO"),
            subtitles: attrs.remove("SUBTITLES"),
            closed_captions: attrs.remove("CLOSED-CAPTIONS"),
        }
    }
```

Am I on the right track or is there an easier way?

Reply:
`QuotedOrNone` only makes sense for the closed caption group id. For other fields you need other enums with their specific special cases. Also I would (for future-proof'ness) do something like `enum ClosedCaptionGroupId { None, Quoted(String), Unquoted(String) }` (or `enum ClosedCaptionGroupId { None, Other(QuotedOrUnquoted) }`).

From the parser you would return a `enum QuotedOrUnquoted { Quoted(String), Unquoted(String) }` and store that in the metadata hashmap, and would then later convert that to the specific types with their special values (like the `NONE` for closed caption group ids) when filling in those struct fields from the metadata hashmap.

Reply:
@sdroege sorry, I'm not following. Yes, it does make sense only for closed caption group-id, but we have the same parser for ANY kv.

Reply:
Unless... I guess I'm a bit slow... 😄 I think i get what you're saying.. let me try this

Reply:
`key_value_pair` would return an `QuotedOrUnquoted`, `from_hashmap` would convert that to the specific types like `ClosedCaptionGroupId` depending on the field that is to be filled in.